ops/pebble.py

```diff
@@ -627,19 +627,26 @@ def wait_change(
         raise TimeoutError(
             'timed out waiting for change {} ({} seconds)'.format(change_id, timeout))
 
-    def add_layer(self, layer: Union[str, dict, Layer]):
-        """Dynamically add a layer to the Pebble setup."""
+    def merge_layer(self, layer: Union[str, dict, Layer]):
+        """Dynamically merge layer onto the Pebble setup layers.
+
+        Pebble merges the layer with the existing dynamic layer according to
+        its flattening/override rules, or adds a new dynamic if there are no
+        dynamic layers.
+        """
         if isinstance(layer, str):
             layer_yaml = layer
```

Comment:
I'd be interested in actually parsing this into a Layer then marshaling back into yaml. The reason for this is the client might change how it sends the layer to the server, to prevent a change in behavior on the client side, enforcing the clients interpretation will ensure compatibility.

Reply:
I'll see what @jameinel  thinks (as I think I originally had the discussion with him), but the idea was to pass through `str` without modification, so that the caller can have full control of what's sent (or to send fields the Operator Framework's `Layer` doesn't yet know about). Though normally charmers will probably call `merge_layer` with a Python `dict` representation, which will go via `Layer` to YAML.

Reply:
I could see going either way with this, but I would want to be consistent.
I think it is reasonable to use Layer as the "we'll be compatible with this" while textual forms wouldn't be guaranteed stable (but would let you use any features you may want before it gets exposed in Layer).